core/src/main/java/dk/alexandra/fresco/framework/network/async/AsyncNetwork.java

```diff
@@ -0,0 +1,303 @@
+            milliSecondsBetween = 500;
+          }
+          int count = 0;
+          boolean connected = false;
+          while (!connected) {
+            try {
+              count++;
+              SocketChannel channel = SocketChannel.open();
+              channel.connect(addr);
+              connected = true;
+              this.clients.put(i, channel);
+            } catch (IOException ex) {
+              if (count > retries) {
+                throw ex;
+              }
+              try {
+                Thread.sleep(milliSecondsBetween);
+              } catch (InterruptedException e) {
+                close();
+                throw new RuntimeException("Got interrupted while trying to connect.");
+              }
+            }
+          }
+        } catch (IOException e) {
+          close();
+          throw new RuntimeException("Could not open a connection to the addr " + addr, e);
+        }
+      }
+    }
+
+    handshake();
+  }
+
+  private class ServerWaiter implements Runnable {
+
+    private SocketChannel channel;
+    private int fromPartyId;
+
+    public ServerWaiter(int fromPartyId, SocketChannel channel) {
+      this.channel = channel;
+      // TODO: Important? Maybe it doesn't matter?
+      // ExceptionConverter.safe(() -> this.channel.configureBlocking(true),
+      // "Could not configure channel to blocking");
+      this.fromPartyId = fromPartyId;
+    }
+
+    @Override
+    public void run() {
+      boolean running = true;
+      while (running) {
+        try {
+          ByteBuffer buf = ByteBuffer.allocate(4);
+          channel.read(buf);
+          buf.flip();
+          int nextMessageSize = buf.getInt();
+          buf = ByteBuffer.allocate(nextMessageSize);
+          while (buf.remaining() > 0) {
+            channel.read(buf);
+          }
+          queues.get(fromPartyId).add(buf.array());
+        } catch (IOException e) {
+          running = false;
+        }
+      }
+    }
+
+  }
+
+  private class Handshaker implements Runnable {
+    private ServerSocketChannel server;
+    private Map<Integer, SocketChannel> partyIdToChannel;
+    private Semaphore blocker;
+    private ConcurrentLinkedDeque<Boolean> successes;
+
+    public Handshaker(ServerSocketChannel server, Map<Integer, SocketChannel> partyIdToChannel,
+        Semaphore blocker, ConcurrentLinkedDeque<Boolean> successes) {
+      this.server = server;
+      this.partyIdToChannel = partyIdToChannel;
+      this.blocker = blocker;
+      this.successes = successes;
+    }
+
+    @Override
+    public void run() {
+      int count = 0;
+      while (count < getNoOfParties() - 1) {
+        try {
+          SocketChannel channel = server.accept();
+          channel.configureBlocking(true);
+          ByteBuffer buf = ByteBuffer.allocate(1);
+          channel.read(buf);
+          buf.position(0);
+          int id = buf.get();
+          partyIdToChannel.put(id, channel);
+          successes.add(true);
```

Comment:
You do not really need this list. Just have a single `boolean` and set it to false if there is an exception. Or am I missing something?